cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -727,6 +729,178 @@ struct CastFunctor<T, DictionaryType,
+
+  if (need_long_long) {
+    long long res;  // NOLINT
+    ibuf >> res;
+    *out = static_cast<T>(res);  // may downcast
+    if (res < min_value || res > max_value) {
+      return false;
+    }
+  } else {
+    long res;  // NOLINT
+    ibuf >> res;
+    *out = static_cast<T>(res);  // may downcast
+    if (res < min_value || res > max_value) {
+      return false;
+    }
+  }
+  return !ibuf.fail() && ibuf.eof();
+}
+
+template <typename T>
+typename std::enable_if<std::is_integral<T>::value && std::is_unsigned<T>::value,
+                        bool>::type static CastStringToNumber(std::istringstream& ibuf,
+                                                              T* out) {
+  static constexpr bool need_long_long = sizeof(T) > sizeof(unsigned long);  // NOLINT
+  static constexpr T max_value = std::numeric_limits<T>::max();
+
+  if (need_long_long) {
+    unsigned long long res;  // NOLINT
+    ibuf >> res;
+    *out = static_cast<T>(res);  // may downcast
+    if (res > max_value) {
+      return false;
+    }
+  } else {
+    unsigned long res;  // NOLINT
+    ibuf >> res;
+    *out = static_cast<T>(res);  // may downcast
+    if (res > max_value) {
+      return false;
+    }
+  }
+  return !ibuf.fail() && ibuf.eof();
+}
+
+template <typename O>
+struct CastFunctor<O, StringType, typename std::enable_if<is_number<O>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    using out_type = typename O::c_type;
+
+    StringArray input_array(input.Copy());
+    auto out_data = GetMutableValues<out_type>(output, 1);
+    errno = 0;
+    // Instantiate the stringstream outside of the loop
+    std::istringstream ibuf;
+    ibuf.imbue(std::locale::classic());
+
+    for (int64_t i = 0; i < input.length; ++i, ++out_data) {
+      if (input_array.IsNull(i)) {
+        continue;
+      }
+      auto str = input_array.GetString(i);
+      ibuf.clear();
+      ibuf.str(str);
+      if (!CastStringToNumber(ibuf, out_data)) {
+        std::stringstream ss;
+        ss << "Failed to cast String '" << str << "' into " << output->type->ToString();
+        ctx->SetStatus(Status(StatusCode::Invalid, ss.str()));
+        return;
+      }
+    }
+  }
+};
+
+// ----------------------------------------------------------------------
+// String to Boolean
+
+// Helper function to cast a C string to a boolean.  Returns true on success,
+// false on error.
+
+static bool CastStringtoBoolean(const char* s, size_t length, bool* out) {
+  if (length == 1) {
+    // "0" or "1"?
+    if (s[0] == '0') {
+      *out = false;
+      return true;
+    }
+    if (s[0] == '1') {
+      *out = true;
+      return true;
+    }
+    return false;
+  }
+  if (length == 4) {
+    // "true"?
+    *out = true;
+    return ((s[0] == 't' || s[0] == 'T') && (s[1] == 'r' || s[1] == 'R') &&
+            (s[2] == 'u' || s[2] == 'U') && (s[3] == 'e' || s[3] == 'E'));
```

Comment:
Can we make this a little more readable by use `tolower`?

Reply:
`tolower` is locale-dependent.